### PR TITLE
Adds magic numbers and versions into bytecode

### PR DESCRIPTION
**Pull request by Yuhanun**
Fixes issue #72 

Format:
4 bytes: 4 random numbers I generated, in this case: 0xB0, 0x45, 0x46, 0xBE
4 bytes: Major version (0 right now)
4 bytes: Minor version (0 right now)
4 bytes: Build version (0 right now)

This would allow us to run separate versions of functions when running the VM for specific versions of bytecode.

Any changes, when requested, will be applied.

Everything has been tested and documented, however no unittests were written.

### CODE REVIEW

**Comment by Yuhanun** [comment]
What version is macOS and windows on and why do they not conform with the standard?

**Comment by Yuhanun** [comment]
Nevermind, forgot to include a header

**Comment by plietar** [comment]
Thanks for tackling this! A few points:
- Could you please update the comment at the top of `bytecode.h`, in the program header section.
- There should be only one definition of the magic bytes. `bytecode.h` is a good place to put that, as that’s where all the shared stuff already is.
- It’s simpler to represent the magic bytes as a 32 bit integer. You can use the functions in Generator and Code to write / read those, rather than manual loops. It’ll be little-endian, but that’s fine as long as the comment around the constant makes that clear.
- I don’t think it’s worth tackling the problem of version numbers yet. We’re very far from a stable release, and we’re not making any kind of guarantees yet. In particular there’s no use in storing those in the Special Descriptors table. You could just drop it, or reduce it to a single `BYTECODE_VERSION` number, defined in one place and enforced at the same time as the magic number when loading the bytecode.

**Comment by Yuhanun** [comment]
Managed to resolve another issue, load<uint32_t> would load a uint16 instead.

File: src/compiler/codegen/codegen.cc [hunk 1]
```diff
@@ -13,6 +13,8 @@
 #include "ds/helpers.h"
 #include "interpreter/bytecode.h"
 
+#include <array>
```
Comment (plietar) [inline src/compiler/codegen/codegen.cc]: Not needed anymore.

File: src/compiler/codegen/codegen.cc [hunk 2]
```diff
@@ -92,6 +94,24 @@ namespace verona::compiler
...
+   * Writes the verona bytes, literally writes VERONA_NUMBEr
+   * @param code Generator to gen from
+   */
+  void write_verona_bytes(Generator& code)
+  {
+    code.u32(bytecode::VERONA_NUMBER);
+  }
+
+  /**
+   * Writes the verona integers to the bytecode
```
Comment (plietar) [inline src/compiler/codegen/codegen.cc]: "Verona integers" is vague. "magic number" would be better.
Also `write_magic_bytes` and `write_verona_bytes` can be one function (eg. `write_magic_number`).

File: src/interpreter/bytecode.h [hunk 3]
```diff
@@ -131,6 +132,8 @@ namespace verona::bytecode
     uint32_t size;
   };
 
+  constexpr static uint32_t VERONA_NUMBER = 4085854915;
```
Comment (plietar) [inline src/interpreter/bytecode.h]: Hex notation is better for this, so it's clear all 4 bytes are used, and it gives something to look for when hexdump'ing a bytecode file.

Should be called `MAGIC_NUMBER` instead, as verona number is too vague.

File: src/interpreter/code.h [hunk 4]
```diff
@@ -5,6 +5,7 @@
 #include "interpreter/bytecode.h"
 #include "interpreter/object.h"
 
+#include <array>
```
Comment (plietar) [inline src/interpreter/code.h]: Not needed anymore

File: src/interpreter/code.h [hunk 5]
```diff
@@ -133,10 +136,19 @@ namespace verona::interpreter
       return header;
     }
 
+    void parse_verona_nums(size_t& ip) {
```
Comment (plietar) [inline src/interpreter/code.h]: This doesn't really parse anything.
How about `check_magic_number` instead? Also this should be a private method.

File: src/interpreter/bytecode.h [hunk 6]
```diff
@@ -13,6 +13,7 @@
  * Bytecode has the following layout:
  *
  * Program header:
+ * - 32-bit verona magic number, 4085854915
```
Comment (plietar) [inline src/interpreter/bytecode.h]: The number should be in hex, or doesn't even have to appear. Just add a "equal to the MAGIC_NUMBER constant" instead.

File: src/compiler/codegen/codegen.cc (hunk 2 above)
Comment (Yuhanun) [reply src/compiler/codegen/codegen.cc]: I kept write_verona_bytes because we might want to add more later, but okay sure

**Comment by orcmid** [comment]
> What version is macOS and windows on and why do they not conform with the standard?

@Yuhanun I think thisissue has been handled.  I just don't understand this question, so I am asking.  What standard?

**Comment by Yuhanun** [comment]
@orcmid The C++ standard, but it was my mistake, I forgot to include and it gave some weird errors on the CI whereas it was fine locally

**Comment by plietar** [comment]
The failure on `features/run-pass/when` is a flaky test, I'll disable and then this LGTM.

**Comment by mjp41** [comment]
@yuhanun I have fixed some issues in the CI process that wasn't enforcing clangformat. Please could you rebase on top of the current master.  The CI is only letting me run without updating the merge point :(

I think you have some clangformat issues.  There is a build target
```
make clangformat
```
or
```
ninja clangformat
```

Should build you a diff to commit to get through the CI.

**Comment by Yuhanun** [comment]
> @Yuhanun I have fixed some issues in the CI process that wasn't enforcing clangformat. Please could you rebase on top of the current master. The CI is only letting me run without updating the merge point :(
> 
> I think you have some clangformat issues. There is a build target
> 
> ```
> make clangformat
> ```
> 
> or
> 
> ```
> ninja clangformat
> ```
> 
> Should build you a diff to commit to get through the CI.

How do I rebase onto it?

**Comment by Yuhanun** [comment]
> @Yuhanun I have fixed some issues in the CI process that wasn't enforcing clangformat. Please could you rebase on top of the current master. The CI is only letting me run without updating the merge point :(
> 
> I think you have some clangformat issues. There is a build target
> 
> ```
> make clangformat
> ```
> 
> or
> 
> ```
> ninja clangformat
> ```
> 
> Should build you a diff to commit to get through the CI.

How do I rebase onto it?

**Comment by Yuhanun** [comment]
Should be good, let me apply linting

**Comment by mjp41** [comment]
Thanks.  For doing this.  I think you have used clang-format 9, but our CI is using an older version.  So they disagree on the style.  I will migrate the CI to a more recent version of clang-format, and then I'll fix any merges for the PR.

**Comment by Yuhanun** [comment]
> 
> 
> Thanks. For doing this. I think you have used clang-format 9, but our CI is using an older version. So they disagree on the style. I will migrate the CI to a more recent version of clang-format, and then I'll fix any merges for the PR.

I'm using clang10, but yeah, cool :) no problem, lmk if anything else can be done by me

**Comment by mjp41** [comment]
I have fixed the CI and created a new PR#81 that incorporates your changes.

**Pull request closed without merging**